lib/charms/prometheus_k8s/v0/prometheus_scrape.py

```diff
@@ -965,18 +1061,28 @@ def alerts(self) -> dict:
             if not alert_rules:
                 continue
 
+            identifier = None
             try:
                 scrape_metadata = json.loads(relation.data[relation.app]["scrape_metadata"])
-                topology = JujuTopology.from_relation_data(scrape_metadata)
-
-                alerts[topology.identifier] = alert_rules
-
+                identifier = ProviderTopology.from_relation_data(scrape_metadata).identifier
+                alerts[identifier] = alert_rules
             except KeyError as e:
-                logger.error(
-                    "Relation %s has invalid data : %s",
+                logger.warning(
+                    "Relation %s has no 'scrape_metadata': %s",
                     relation.id,
                     e,
                 )
+                logger.debug("Without scrape_metadata, check for an aggregate or forwarder")
+                if "groups" in alert_rules["groups"][0].keys():
```

Comment:
I appreciate that this if clause is being added to fix a bug. But this make the code difficult to understand. I wonder if we can do some refactoring here to remove the problem of
1. groups nested in groups
2. need to add topology to rule name only to remove it again. 

Also in view of limited time we can postpone this refactoring by creating a ticket, since it seems there may be no change to the public interface when we do refactor. But double check me on this.

Reply:
I went looking for your fix today to merge it it in here, but didn't find it. What was the resolution?

We definitely agree here that this is difficult to understand, and it would be ideal to kill the nesting. We do not add topology to the rule name only to remove it again, though -- instead, we add topology to the rule name in the nested groups, those are directly yanked out by the regex, and a new dict is created which matches the "expected": `alerts[identifier] = alert_rules` syntax.

The real downside to it is that it's a regular expression.

Reply:
If as I understand this change was to fix the bug that resulted from nested groups (`{"groups" : {"groups": ...}}`) then that was a bug in the scrape config charm. A fix is available in [this branch](https://github.com/canonical/prometheus-scrape-config-k8s-operator/commits/openg-528) of the scrape config charm.